cs/lexer.adoc

```diff
@@ -0,0 +1,271 @@
+
+* datový typ první hodnoty ve vstupu (datatype!)
+
+* zbývající část vstupu za skenovanou hodnotou (binary! string!)
+
+
+== Instrumentace lexeru
+
+=== Schema práce lexeru
+
+Proces tokenizace je rozdělen do jednotlivých stupňů, v nichž jsou spouštěny události tam, kde může být invokována uživatelem poskytnutá funkce se zpětným voláním (callback). Jednotlivými stupni jsou:
+
+----
+                +-> ERROR
+               /
+         +-> CLOSE series
+        /
+       +-> OPEN series
+      /
+-> PRESCAN token -> SCAN token -> LOAD value
+      \               \             \
+       +-> ERROR       +-> ERROR     +-> ERROR
+----
+
+Událostmi lexeru tedy jsou: `prescan`, `scan`, `load`, `open`, `close`, `error`.
+
+=== `transcode/trace`
+
+*Syntaxe*
+
+----
+transcode/trace <input> <callback>
+
+<input>    : načítaná řada (binary! string!).
+<callback> : zpětná (callback) funkce pro ošetření události lexeru (function!).
+
+Vrací    : jedinou hodnotu nebo blok hodnot.
+----
+
+*Popis*
+
+Konvertuje vstupní binární řadu v kódování UTF-8 na interně použitelné hodnoty ve shodě se syntaktickými pravidly. Při každé události lexeru volá uživatelem poskytnutou callback funkci.
+
+Blok specifikací callback funkce:
+
+----
+func [
+    event [word!]               ;-- aktuální stav lexeru (viz tabulka níže)
+    input [string! binary!]     ;-- odkaz na vstupní řadu v aktuální pozici načítání (může být změněno)
+    type  [datatype! word!]     ;-- slovo nebo datový typ popisující typ tokenu nebo aktuálně řešenou hodnotu
+    line  [integer!]            ;-- číslo řady aktuálního vstupu
+    token                       ;-- aktuální token jako úsek (slice) vstupu  typu pair! nebo načítaná hodnota
+    return: [logic!]
+][
+    [events]                    ;-- volitelný seznam vymezených názvů událostí
+    ...body...
+]
+----
+Odsazení (offset) argumentu `vstupní` řady je dáno místem, kde se lexer zastavil po detekci konce tokenu. Tento offset může být modifikován callback funkcí. Je-li `chybová` událost ignorována, nedojde automaticky k pokročení vstupem; záleží na callback funkci aby nastavila `vstupní` řadu do správné pozice pro opětovné ošetření řady. Pokud se nezadaří, může to vést k nekonečné smyčce.
+
+Při `chybové` události obsahuje argument `type` název datového typu, jenž byl částečně rozpoznán. Vyskytne-li se chyba u izolovaného znaku, jako jsou nepárové koncové závorky `) ] }`, je argument `type` nastaven na `error!`, protože v takovém případě nebyl rozpoznán žádný určitý typ.
+
+Blok těla funkce může začínat volitelným filtrovacím blokem k určení události, která má být spuštěna. To umožňuje redukovat počet zpětných (callbac) volání, což se projeví lepším výkonem zpracování.
+
+Význam některých argumentů a zpětných hodnot __závisí__ na typu události. Následující tabulka obsahuje možné kombinace a účinky:
+
+[cols="1,1,1,1,2", options="header"]
+|===
+|Event | Type | Token | Return Value | Description
```

Comment:
Header is not translated ("event", "type", "token" etc).

Reply:
Its partly a lazines and partly a didactical intention:  force a rader to learn english words.
Can be changed.